commit 012d4869c1eb195e83f159ed7b2bced33f37f960
Author: Stefan Hajnoczi <stefanha@linux.vnet.ibm.com>
Date:   Wed Apr 21 20:35:45 2010 +0100

    block: Free iovec arrays allocated by multiwrite_merge()
    
    A new iovec array is allocated when creating a merged write request.
    This patch ensures that the iovec array is deleted in addition to its
    qiov owner.
    
    Reported-by: Leszek Urbanski <tygrys@moo.pl>
    Signed-off-by: Stefan Hajnoczi <stefanha@linux.vnet.ibm.com>
    Signed-off-by: Kevin Wolf <kwolf@redhat.com>
    (cherry picked from commit 1e1ea48d42e011b9bdd0d689d184e7cac4617b66)

Origin: upstream, http://git.kernel.org/?p=virt/kvm/qemu-kvm.git;a=commit;h=012d4869c1eb195e83f159ed7b2bced33f37f960
Bug-ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu-kvm/+bug/588293

diff --git a/block.c b/block.c
index 4f9a48b..7326bfe 100644
--- a/block.c
+++ b/block.c
@@ -1608,6 +1608,9 @@ static void multiwrite_user_cb(MultiwriteCB *mcb)
 
     for (i = 0; i < mcb->num_callbacks; i++) {
         mcb->callbacks[i].cb(mcb->callbacks[i].opaque, mcb->error);
+        if (mcb->callbacks[i].free_qiov) {
+            qemu_iovec_destroy(mcb->callbacks[i].free_qiov);
+        }
         qemu_free(mcb->callbacks[i].free_qiov);
         qemu_vfree(mcb->callbacks[i].free_buf);
     }
